chapters/annotations.tex

```diff
@@ -2052,24 +2055,28 @@ \subsection{Protection of Classes}\label{protection-of-classes}
   Same as \lstinline!Access.packageText! and additionally the class, or part of the class, can be copied.
 \end{enumerate}
 
-The \lstinline!access! annotation holds for the respective class and all classes
-that are hierarchically on a lower level, unless overridden by a
-\lstinline!Protection! annotation with \lstinline!access!.
-Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate!
-has no effect.
+The \lstinline!access! annotation holds for the respective class and all classes that are hierarchically on a lower level, unless overridden by a \lstinline!Protection! annotation with \lstinline!access!.
+Overriding \lstinline!access=Access.hide! or \lstinline!access=Access.packageDuplicate! has no effect.
+
+Classes should not use other classes in ways that contradict this protection.
+Tools must ensure that protected contents are not shown, even if classes do not meet this requirement.
```

Comment:
Stronger:
```suggestion
A class may only be used in ways that comply with this protection.
Tools must ensure that protected contents are not revealed, even in case this rule about valid use is violated.
```

Reply:
The original text used should and must according to https://www.rfc-editor.org/rfc/rfc2119 - "may not" is not included in that list. (Well, "should" was possibly borderline.) Thus, I prefer to not change this. The other words could be changed.

Reply:
It is this very meaning from RFC2119 that I find inappropriate in these statements:
> **[4]. SHOULD NOT**   This phrase, or the phrase "NOT RECOMMENDED" mean that there may exist valid reasons in particular circumstances when the particular behavior is acceptable or even useful, but the full implications should be understood and the case carefully weighed before implementing any behavior described with this label.

I don't think that the specification should open up for using the classes in ways that contradict the protection.  This way, the responsibility of the library developer is clear, but the second sentence still adds extra security for the library developer in case he/she is using a tool that fails to verify the protection requirements when preparing the library for distribution.

Reply:
> It is this very meaning from RFC2119 that I find inappropriate in these statements:
> 
> > **[4]. SHOULD NOT**   This phrase, or the phrase "NOT RECOMMENDED" mean that there may exist valid reasons in particular circumstances when the particular behavior is acceptable or even useful, but the full implications should be understood and the case carefully weighed before implementing any behavior described with this label.
> 
> I don't think that the specification should open up for using the classes in ways that contradict the protection. This way, the responsibility of the library developer is clear, but the second sentence still adds extra security for the library developer in case he/she is using a tool that fails to verify the protection requirements when preparing the library for distribution.

If it is an **absolute requirement**, I don't see that it makes sense to specify the behavior.
We could clarify that "The only acceptable use of class contradicting the protection is for testing the protection."

Reply:
I agree that it would be best to not specify the behavior.  I would prefer to not clarify anything at all:
- The specification shouldn't say anything at all regarding what kind of testing tool vendors should perform in order to check their compliance with the specification.  On GitHub, we _could_ provide a test collection that can help tool vendors check their compliance, but it is not the specification's job to point out such resources.
- Instead of hinting that library authors perhaps should worry about how tools behave if the library violates the specification, we should explain the importance of library authors verifying that a tool meets their expectations when it comes to supporting correct `protection`-annotation behavior (for valid cases), before they allow their library to be decrypted by the tool.

Reply:
> I agree that it would be best to not specify the behavior.

I don't. The reason is that such issues occur does occur in practice.
However, if we don't want to specify the behavior it seems odd to make it an absolute requirement.

Reply:
One point is at _what level_ it shall be rejected.

If you encrypt library of 100 models, does a tool have to do some detailed(*) check that _all of them_ satisfy this before doing anything?

I would say not - and thus I see two realistic alternatives:
- Restrict what is shown for that model, but show other models
- Don't show the contents of that model, but show other models

Both are ok to me; but neither is consistent with viewing it as an absolute requirement.
Note that this may get complicated if the problematic model is used internally in the library.

*: There are also some non-detailed checks. E.g., if a tool cannot parse the library it doesn't make sense to present anything. However, as far as I understand SEMLA is designed so that even parsing of sub-parts can be delayed (similarly as for normal Modelica).

Reply:
This was resolved at the latest phone meeting.